Merge heh and zhehuai's initial version of frozen am+llm

zhehuaichen opened the pull request:
# What does this PR do ?

This PR tries to merge heh and zhehuai's implementation of frozen am+llm. These implementations will contribute to *Modular Speech LLM​* [slide](https://nvidia.sharepoint.com/:p:/r/sites/DLAlgorithmResearchTeamSite/Presentations/Roadmap/Nemo_SpeechLM_H2_2023.pptx?d=w6e870786c30a4bf6a65cbf9ed08c444a&csf=1&web=1&e=eq35C4&nav=eyJzSWQiOjU1MDE0NTQ1MCwiY0lkIjoyNjIxMjQxMzV9) idea the SpeechLLM team plans to pursue.
 
heh's initial version: https://github.com/stevehuang52/NeMo/commit/36a1fbfad456d8c19d7fb3b3fbd181aadc0cb466
Zhehuai's initial version: https://github.com/zhehuaichen/NeMo/compare/main...zhehuaichen:NeMo:speechllm
Their previous differences are summarized here and will be resolved in this PR and its followups:
https://docs.google.com/document/d/1zNI4hC6vJtUfcHbrUSPaMuYWRBQdN_36H0P2NiBiuPY/edit

This PR includes
1. Finish merging the model and dataset code
2. Previous tests are still enabled and passed (prepare_llm_input, training_step, validation_step)

The major remaining works are listed here
https://docs.google.com/document/d/1o0AM7v4gcTQkPZjE0Vl9TTX4vYnGTrbXEFGWh0UhGlk/edit#bookmark=id.pzvdadt5oxyw

zhehuaichen committed:
Merge heh and zhehuai's initial version of frozen am+llm

The previous differences are summarized here:
https://docs.google.com/document/d/1zNI4hC6vJtUfcHbrUSPaMuYWRBQdN_36H0P2NiBiuPY/edit

This PR includes
1. Finish merging the model and dataset code
2. Previous tests are still enabled and passed (prepare_llm_input, training_step,
    validation_step)

The major remaining works are listed here
https://docs.google.com/document/d/1o0AM7v4gcTQkPZjE0Vl9TTX4vYnGTrbXEFGWh0UhGlk/edit#bookmark=id.pzvdadt5oxyw

stevehuang52 committed:
clean up modularized_speech_gpt_config.yaml

Signed-off-by: He Huang (Steve) <105218074+stevehuang52@users.noreply.github.com>

stevehuang52 reviewed (changes requested):
Thanks for the great work, the PR is almost ready to merge, just missing the few things that we discussed yesterday.
1) change `init_perception_model` to make it properly instantiate all perception sub-modules
2) change the yaml config to make it use the new perception modules
3) run the example training script with LS960 to make sure the training pipeline works

zhehuaichen committed:
1. change init_perception_model to make it properly instantiate all
perception sub-modules
2. change the yaml config to make it use the new perception modules
3. run the example training script with LS960 to make sure the training
pipeline works
4. also fix the python format

zhehuaichen committed:
Merge branch 'speechllm_merge' of github.com:zhehuaichen/NeMo into speechllm_merge

stevehuang52 reviewed (approved):
LGTM, thanks~
